File: sample/src/main/java/com/mateuszkoslacz/moviper/sample/viper/interactor/ListingInteractor.java
package com.mateuszkoslacz.moviper.sample.viper.interactor;

import com.mateuszkoslacz.moviper.base.interactor.BaseInteractor;
import com.mateuszkoslacz.moviper.sample.data.retrofit.GitHubApiInterface;
import com.mateuszkoslacz.moviper.sample.viper.contract.ListingContract;

import java.io.IOException;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ListingInteractor
        extends BaseInteractor<ListingContract.PresenterForInteractor>
        implements ListingContract.Interactor {

    private GitHubApiInterface mGitHubApiInterface;

    public ListingInteractor() {
        final Retrofit mRetrofit = new Retrofit.Builder()
                .baseUrl(GitHubApiInterface.GITHUB_API_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        mGitHubApiInterface = mRetrofit.create(GitHubApiInterface.class);
    }

    @Override
    public void getUserList() {
        if (isPresenterAttached()) {
            Thread thread = new Thread(() -> {
                try {
                    getPresenter().onUserFetched(mGitHubApiInterface
                            .getUsers().execute().body());
                } catch (IOException e) {
                    getPresenter().onUserFetchedError(e);
                }
            });

            thread.start();
        }
    }
}

File: sample-super-rx-ai-kotlin/src/main/java/com/mateuszkoslacz/moviper/rxsample/viper/interactor/ListingInteractor.kt
package com.mateuszkoslacz.moviper.rxsample.viper.interactor

import com.jakewharton.retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import com.mateuszkoslacz.moviper.base.interactor.BaseRxInteractor
import com.mateuszkoslacz.moviper.rxsample.data.retrofit.GitHubApiInterface
import com.mateuszkoslacz.moviper.rxsample.viper.contract.ListingContract
import com.mateuszkoslacz.moviper.rxsample.viper.entity.User
import io.reactivex.Observable
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

open class ListingInteractor : BaseRxInteractor(), ListingContract.Interactor {

    private val mGitHubApiInterface: GitHubApiInterface

    init {
        val mRetrofit = Retrofit.Builder()
                .baseUrl(GitHubApiInterface.GITHUB_API_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build()
        mGitHubApiInterface = mRetrofit.create(GitHubApiInterface::class.java)
    }

    override val userList: Observable<List<User>>
        get() = mGitHubApiInterface.users
}
